middle_end/flambda2/tests/lib/minicheck/type.ml

```diff
@@ -0,0 +1,270 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*           Mark Shinwell and Leo White, Jane Street Europe              *)
+(*                                                                        *)
+(*   Copyright 2015--2022 Jane Street Group LLC                           *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+type 'a t =
+  { generate : Splittable_random.t -> 'a;
+    print : Format.formatter -> 'a -> unit
+  }
+
+let generate t = t.generate
+
+let print t = t.print
+
+let print_unknown ppf _ = Format.pp_print_string ppf "_"
+
+let define ~generate ?(print = print_unknown) () = { generate; print }
+
+let with_print t ~print = { t with print }
+
+let bool =
+  let generate r = Splittable_random.int r mod 2 = 0 in
+
+  define () ~generate ~print:Format.pp_print_bool
+
+let int = define () ~generate:Splittable_random.int ~print:Format.pp_print_int
+
+let map_generate ty ~f =
+  let generate r = f (generate ty r) in
+  define () ~generate ~print:(print ty)
+
+(* Should be small because we're doing this the dumb way *)
+let small_nat ~less_than =
+  map_generate int ~f:(fun i -> Int.abs i mod less_than)
+
+let log_int =
+  let generate r =
+    (* This assumes that the size of [int] is one less than the word size, which
+       isn't true for JavaScript *)
+    let bits = Splittable_random.int r mod Sys.word_size in
+    if bits = 0
+    then 0
+    else
+      let high_bit = 1 lsl (bits - 1) in
+      let mask = high_bit - 1 in
+      Splittable_random.int r land mask lor high_bit
+  in
+  define () ~generate ~print:Format.pp_print_int
```

Comment:
Picking an int according to a logarithmic distribution (or log-uniform, I guess?). In `Splittable_random` they put this and others in a module called `Log_uniform` \*shrug\*

I agree this code could be clearer, though ...

Reply:
Aha, I hadn't encountered that term before.  [Wikipedia provides an outline.](https://en.wikipedia.org/wiki/Reciprocal_distribution)  I guess that helps smooth things out in terms of generating a wide variety of bit strings?

Reply:
Right. Also it makes collisions and near-collisions slightly more likely. (Though now that I think of it it's probably not _that_ important since the high bits are likely to be ignored anyway.)

Reply:
Yeah, this sounds great, now that I understand it.